src/backend/distributed/transaction/relation_access_tracking.c

```diff
@@ -434,6 +500,78 @@ RecordParallelRelationAccessToCache(Oid relationId,
 }
 
 
+/*
+ * TrySettingLocalMultiShardModifyModeToSequential is a wrapper around
+ * SetLocalMultiShardModifyModeToSequential() with an error check.
+ *
+ * The error check ensures that setting the mode to sequential doesn't
+ * contradict with our purpose of using a single connection per worker
+ * node. Note that using a single conneection per worker is crucial for
+ * implementing foreign keys to reference tables, where reference tables
+ * is likely to CASCADE (and acquire locks) to multiple placements
+ * of a distributed table over a single connection.
+ */
+static void
+TrySettingLocalMultiShardModifyModeToSequential(void)
+{
+	if (ParallelQueryExecutedInTransaction())
+	{
+		/*
+		 * If there has already been a parallel query executed, the sequential mode
+		 * would still use the already opened parallel connections to the workers,
+		 * thus contradicting our purpose of using sequential mode.
+		 */
+		ereport(ERROR, (errmsg("attempt to set query execution mode to sequential "
+							   "failed since a parallel query has already been "
+							   "executed in the same transaction"),
+						errhint("Try re-running the transaction with "
+								"\"SET LOCAL citus.multi_shard_modify_mode TO "
+								"\'sequential\';\"")));
+	}
+
+	SetLocalMultiShardModifyModeToSequential();
+}
+
+
+/*
+ * ParallelQueryExecutedInTransaction returns true if any parallel query
+ * is executed in the current transaction.
+ */
+bool
+ParallelQueryExecutedInTransaction(void)
+{
+	HASH_SEQ_STATUS status;
+	RelationAccessHashEntry *hashEntry;
+	int smallestParallelRelationAccessBit = 0;
+	int smallestParallelValue = 0;
+
+	if (!ShouldRecordRelationAccess() || RelationAccessHash == NULL)
+	{
+		return false;
+	}
+
+	smallestParallelRelationAccessBit =
+		PLACEMENT_ACCESS_SELECT + PARALLEL_MODE_FLAG_OFFSET;
+	smallestParallelValue = (1 << smallestParallelRelationAccessBit);
+
+	hash_seq_init(&status, RelationAccessHash);
+
+	hashEntry = (RelationAccessHashEntry *) hash_seq_search(&status);
+	while (hashEntry != NULL)
+	{
+		if (hashEntry->relationAccessMode >= smallestParallelValue)
```

Comment:
This seems slightly dodgy, e.g. it breaks if we start using additional bits with a different meaning. 

Maybe we should have a bit mask `(hashEntry->relationAccessMode & PARALLEL_ACCESS_MASK) != 0`